linkage-monitor/src/main/java/com/google/cloud/tools/dependencies/linkagemonitor/Analytics.java

```diff
@@ -0,0 +1,223 @@
+   * ultimately funnels through here. 
+   */
+  private void ping(PingEvent pingEvent) {
+    if (sendAnalytics && pingEvent != null) {
+      try {
+        String json = jsonEncode(pingEvent);
+        int resultCode = HttpUtil.sendPost(collectionUrl, json, "application/json");
+        if (resultCode >= 300) {
+          logger.log(Level.FINE, "Failed to POST to Concord with HTTP return code " + resultCode);
+        }
+      } catch (IOException ex) {
+        // Don't recover or retry.
+        logger.log(Level.FINE, "Failed to POST to Concord", ex);
+      } 
+    }
+  }
+
+  @VisibleForTesting
+  static class PingEvent {
+    @VisibleForTesting final String eventName;
+    @VisibleForTesting final ImmutableMap<String, String> metadata;
+
+    PingEvent(String eventName, ImmutableMap<String, String> metadata) {
+      Preconditions.checkArgument(!Strings.isNullOrEmpty(eventName), "eventName null or empty");
+      Preconditions.checkNotNull(metadata, "metadata is null");
+      this.eventName = eventName;
+      this.metadata = metadata;
+    }
+  }
+
+  @VisibleForTesting
+  String jsonEncode(PingEvent event) {
+    Gson gson = new Gson();
+
+    Map<String, String> desktopClientInfo = new HashMap<>();
+    desktopClientInfo.put("os", System.getProperty("os.name"));
+    
+    Map<String, Object> clientInfo = new HashMap<>();
+    clientInfo.put("client_type", "DESKTOP");
+    clientInfo.put("desktop_client_info", desktopClientInfo);
+        
+    // logs/proto/cloud/concord/concord_event.proto
+    Map<String, Object> sourceExtension = new HashMap<>();
+    sourceExtension.put("client_install_id", getAnonymizedClientId());
+    sourceExtension.put("console_type", CONSOLE_TYPE);
+    sourceExtension.put("event_name", event.eventName);
+    
+    Map<String, String> metadataMap = new LinkedHashMap<>(event.metadata);
+
+    List<Map<String, String>> metadataList = new ArrayList<>();
+    for (Map.Entry<String, String> entry : metadataMap.entrySet()) {
+      Map<String, String> keyValue = new HashMap<>();
+      keyValue.put("key", entry.getKey());
+      keyValue.put("value", entry.getValue());
+      metadataList.add(keyValue);
+    }
+    sourceExtension.put("event_metadata", metadataList);
+    
+    String sourceExtensionJsonString = gson.toJson(sourceExtension);
+    
+    Map<String, Object> logEvent = new HashMap<>();
+    logEvent.put("event_time_ms", System.currentTimeMillis());
+    logEvent.put("sequence_position", sequencePosition++);  
+    logEvent.put("source_extension_json", sourceExtensionJsonString);
+    
+    Map<String, Object> root = new HashMap<>();
+    root.put("log_source", "CONCORD");
+    root.put("request_time_ms", System.currentTimeMillis());
+    root.put("client_info", clientInfo);
+    root.put("log_event", Collections.singletonList(logEvent));
+
+    return gson.toJson(root);
+  }
+
+  @VisibleForTesting
+  static String getAnonymizedClientId() {
+    return "clientId"; // TODO fill in with what????";
```

Comment:
I don't recall a specific and exact requirement for "client_install_id", but basically, in CT4E, this is for distinguishing and counting different "users" or "clients." In CT4E, we just generated a UUID that will be persisted in an Eclipse instance, hopefully matching one unique "user" per Eclipse installation (or instance?).